Remix Vite plugin support

TooTallNate opened the pull request:
Adds support for Remix apps which use the new Remix Vite plugin.

* The vanilla Remix + Vite template deploys correctly out-of-the-box, however only a single Node.js function will be used, and a warning will be printed saying to configure the `vercelPreset()` Preset.
* When used in conjunction with the `vercelPreset()` Preset (https://github.com/vercel/remix/pull/81), allows for the application to utilize Vercel-specific features, like per-route `export const config` configuration, including multi-runtime (Node.js / Edge runtimes) within the same app.

## To test this today

1. Generate a Remix + Vite project from the template:
    ```
    npx create-remix@latest --template remix-run/remix/templates/vite
    ```
1. Install `@vercel/remix`:
    ```
    npm i --save-dev @vercel/remix
    ```
1. Ensure you are using Remix v2.8.0 or later.
1. Configure the `vercelPreset()` in the `vite.config.ts` file:
    ```diff
    --- a/vite.config.ts
    +++ b/vite.config.ts
    @@ -1,10 +1,11 @@
     import { vitePlugin as remix } from "@remix-run/dev";
     import { installGlobals } from "@remix-run/node";
     import { defineConfig } from "vite";
    +import { vercelPreset } from "@vercel/remix/vite";
     import tsconfigPaths from "vite-tsconfig-paths";
    
     installGlobals();
    
     export default defineConfig({
    -  plugins: [remix(), tsconfigPaths()],
    +  plugins: [remix({ presets: [vercelPreset()] }), tsconfigPaths()],
     });
    ```
1. Create a new Vercel Project in the dashboard, and ensure the Vercel preset is set to "Remix" in the Project Settings. The autodetection will work correctly once this PR is merged, but for now it gets incorrectly detected as "Vite" preset.
    * **Hint**: You can create a new empty Project by running the `vercel link` command.
    <img width="545" alt="Screenshot 2024-02-27 at 10 37 11" src="https://github.com/vercel/vercel/assets/71256/f46baf57-5d97-4bde-9529-c9165632cb30">
1. Deploy to Vercel:
    ```
    vercel deploy
    ```

TooTallNate committed:
Make "remix" framework preset supersede "vite"

TooTallNate committed:
[WIP] Remix Vite plugin support

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into tootallnate/zero-1217-research-remix-v-next-vite

TooTallNate committed:
.

TooTallNate committed:
Progress point

TooTallNate committed:
Remove some `console.log`

TooTallNate committed:
.

TooTallNate committed:
Add default handling when not using Preset

TooTallNate committed:
Add new e2e test

TooTallNate committed:
.

TooTallNate committed:
Add probes

TooTallNate committed:
.

TooTallNate committed:
Merge branch 'main' of github.com:vercel/vercel into tootallnate/zero-1217-research-remix-v-next-vite

TooTallNate committed:
Add fixture

TooTallNate committed:
Add changeset

TooTallNate committed:
.

TooTallNate commented on packages/remix/test/fixtures-vite/01-no-preset/vercel.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "framework": "remix"
```
Remix + Vite is currently detected as `vite` preset. The changes in this PR fix that, but these changes need to be deployed on the API-side first. So we can remove these `vercel.json` files in the fixtures in a follow-up PR after this PR is deployed.

TooTallNate committed:
Ensure < 12 functions

TooTallNate committed:
.

TooTallNate committed:
Fix assets cache-control header

TooTallNate commented on packages/remix/test/fixtures-vite/02-interactive-remix-routing-v2/package.json:
```diff
@@ -0,0 +1,43 @@
+    "react-dom": "^18.2.0"
+  },
+  "devDependencies": {
+    "@remix-run/dev": "https://remix-e7asp269z.vercel.sh/remix-dev.tgz",
```
**Note:** We'll want to update this to 2.7.3 once that version is released.

TooTallNate committed:
Support custom `assetsDir`

TooTallNate committed:
Merge branch 'main' into tootallnate/zero-1217-research-remix-v-next-vite

depsimon commented:
> * The vanilla Remix + Vite template deploys correctly out-of-the-box

Hi @TooTallNate can you point out the vanilla Remix + Vite template you used?

I tried with `create-remix@latest --template remix-run/remix/templates/vite` but this does not deploy correctly out-of-the-box on Vercel (deploys correctly but getting 404 after configuring the output directory to `/build/client`).

Did you use a specific config to make it work?

k1sul1 commented:
Great work on this! Any chance of getting an ETA on when this could be available in production? 

Next week? Next month? Any info would make it much easier to decide on how we should proceed with a blocked PR and will be greatly appreciated!

TooTallNate committed:
Add framework detection unit test

TooTallNate committed:
Another unit test

ToppleTheNun commented:
After following the instructions in the PR body, I'm getting these errors on my test repo:
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/var/task/build/index.js' imported from /var/task/build/server-index.mjs
    at new NodeError (node:internal/errors:405:5)
    at finalizeResolution (node:internal/modules/esm/resolve:332:11)
    at moduleResolve (node:internal/modules/esm/resolve:1029:10)
    at moduleResolveWithNodePath (node:internal/modules/esm/resolve:876:12)
    at defaultResolve (node:internal/modules/esm/resolve:1242:79)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:403:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:372:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:249:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:39)
    at link (node:internal/modules/esm/module_job:75:36) {
  url: 'file:///var/task/build/index.js',
  code: 'ERR_MODULE_NOT_FOUND'
}
INIT_REPORT Init Duration: 325.29 ms	Phase: invoke	Status: error	Error Type: Runtime.ExitError
Error: Runtime exited with error: exit status 1
```

Minimal reproduction: https://github.com/ToppleTheNun/remix-vite-vercel

TooTallNate commented:
@ToppleTheNun Could you share the deployment URL? It deployed correctly for me: https://remix-vite-vercel-epvyuy9p4-uncurated-tests.vercel.app/

ToppleTheNun commented:
@TooTallNate https://vercel.com/topplethenun/remix-vite-vercel/G8UPjGaGrdPJ4FA8uytXBWy8RDp2

TooTallNate commented:
Looks like you missed the `VERCEL_CLI_VERSION` env var part, so it's using the wrong version of the CLI which doesn't contain the changes in this PR.

ToppleTheNun commented:
I ran the deploy from my local with this:
```shell
$ vercel deploy -b VERCEL_CLI_VERSION=https://vercel-git-tootallnate-zero-1217-research-remix-v-next-vite.vercel.sh/tarballs/vercel.tgz
```

ToppleTheNun commented:
Ah, I see where it went wrong. I clicked the "promote to production" link in the Vercel console and it built without using the CLI. That'd do it. Apologies!

TooTallNate commented:
<img width="1101" alt="Screenshot 2024-02-27 at 14 44 15" src="https://github.com/vercel/vercel/assets/71256/a9c2f41a-99a9-4eca-bc71-93ed2975254e">

You should see this log line in the deployment build logs (which I don't see for your deployment). So looks like it didn't take for some reason. Perhaps try setting the env var in the Project Settings on the Vercel dashboard instead.

justyn-clark commented:
<img width="1075" alt="Screenshot 2024-02-27 at 2 34 27 PM" src="https://github.com/vercel/vercel/assets/11221796/5a3eaed4-bbc1-466a-b3cf-a7a9033cd49a">


<img width="853" alt="Screenshot 2024-02-27 at 2 39 38 PM" src="https://github.com/vercel/vercel/assets/11221796/aa152a4e-79c4-4748-9741-e5bff2735f3b">

------------

This deployment was successful pushing the branch to Github
https://remix-vite-vercel-test-dvfhsdxv6-justyn-clark.vercel.app

This deployment was successful using vercel deploy from step 6. above
https://remix-vite-vercel-test-bg8ziwstj-justyn-clark.vercel.app

```
vercel deploy -b VERCEL_CLI_VERSION=https://vercel-git-tootallnate-zero-1217-research-remix-v-next-vite.vercel.sh/tarballs/vercel.tgz
```

------------

Out of the box I'm a happy camper. Nice job and thanks for giving this attention.  🎉

TooTallNate committed:
Invert Vite vs. legacy check

It's possible to specify a different file name for the Vite config, so
it's better to check for the legacy remix config file name instead.

depsimon commented:
> @depsimon I've added some instructions at the top. Let me know if you bump into any issues.
> 
> @k1sul1 We're aiming for this week. Remix v2.8.0 also needs to be released as it contains a bug fix related to the server bundles feature (but I believe that is happening in the next couple of days).

Thanks @TooTallNate, works fine following your instructions 🦾

krall12 commented:
Following the steps with the custom VERCEL_CLI_VERSION I'm getting a bun install error. 

<img width="1250" alt="Screenshot 2024-02-28 at 10 57 57 AM" src="https://github.com/vercel/vercel/assets/17836325/f36d3a12-e55a-498b-aa46-1c5fa5e8e10a">

EndangeredMassa reviewed (approved):
<!-- Suggestion comments propose improvements to the code. Mention specifically what should be changed and what the benefit would be.  -->
**suggestion (soft-blocking):** add unit tests

The logic overall seems good, but there's a lot of it, making it harder to review as a whole. I think we should break more of that into testable chunks at the unit level.
